builtin/providers/aws/resource_aws_emr_cluster.go

```diff
@@ -0,0 +1,537 @@
+		emrSlaveSecGrp = attributes["emr_managed_slave_security_group"].(string)
+
+		if len(strings.TrimSpace(attributes["instance_profile"].(string))) != 0 {
+			instanceProfile = strings.TrimSpace(attributes["instance_profile"].(string))
+		}
+	}
+
+	emrApps := expandApplications(applications)
+
+	params := &emr.RunJobFlowInput{
+		Instances: &emr.JobFlowInstancesConfig{
+			Ec2KeyName:                  aws.String(userKey),
+			Ec2SubnetId:                 aws.String(subnet),
+			InstanceCount:               aws.Int64(int64(coreInstanceCount + 1)),
+			KeepJobFlowAliveWhenNoSteps: aws.Bool(true),
+			MasterInstanceType:          aws.String(masterInstanceType),
+			SlaveInstanceType:           aws.String(coreInstanceType),
+			TerminationProtected:        aws.Bool(false),
+			AdditionalMasterSecurityGroups: []*string{
+				aws.String(extraMasterSecGrp),
+			},
+			AdditionalSlaveSecurityGroups: []*string{
+				aws.String(extraSlaveSecGrp),
+			},
+			EmrManagedMasterSecurityGroup: aws.String(emrMasterSecGrp),
+			EmrManagedSlaveSecurityGroup:  aws.String(emrSlaveSecGrp),
+		},
+		Name:         aws.String(d.Get("name").(string)),
+		Applications: emrApps,
+
+		ReleaseLabel:      aws.String(d.Get("release_label").(string)),
+		ServiceRole:       aws.String(d.Get("service_role").(string)),
+		VisibleToAllUsers: aws.Bool(d.Get("visible_to_all_users").(bool)),
+	}
+
+	if instanceProfile != "" {
+		params.JobFlowRole = aws.String(instanceProfile)
+	}
+
+	if v, ok := d.GetOk("log_uri"); ok {
+		logUrl := v.(string)
+		params.LogUri = aws.String(logUrl)
+	}
+	if v, ok := d.GetOk("bootstrap_action"); ok {
+		bootstrapActions := v.(*schema.Set).List()
+		params.BootstrapActions = expandBootstrapActions(bootstrapActions)
+	}
+	if v, ok := d.GetOk("tags"); ok {
+		tagsIn := v.(map[string]interface{})
+		params.Tags = expandTags(tagsIn)
+	}
+	if v, ok := d.GetOk("configurations"); ok {
+		confUrl := v.(string)
+		params.Configurations = expandConfigures(confUrl)
+	}
+
+	log.Printf("[DEBUG] EMR Cluster create options: %s", params)
+	resp, err := conn.RunJobFlow(params)
+
+	if err != nil {
+		log.Printf("[ERROR] %s", err)
+		return err
+	}
+
+	log.Printf("[DEBUG] Created EMR Cluster done...")
+	d.SetId(*resp.JobFlowId)
+
+	log.Println(
+		"[INFO] Waiting for EMR Cluster to be available")
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"STARTING", "BOOTSTRAPPING"},
+		Target:     []string{"WAITING", "RUNNING"},
+		Refresh:    resourceAwsEMRClusterStateRefreshFunc(d, meta),
+		Timeout:    40 * time.Minute,
+		MinTimeout: 10 * time.Second,
+		Delay:      30 * time.Second, // Wait 30 secs before starting
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("[WARN] Error waiting for EMR Cluster state to be \"WAITING\": %s", err)
+	}
+
+	return resourceAwsEMRClusterRead(d, meta)
+}
+
+func resourceAwsEMRClusterRead(d *schema.ResourceData, meta interface{}) error {
+	emrconn := meta.(*AWSClient).emrconn
```

Comment:
We are not setting many of the params back to state in this Read func - for future proofing, this will cause issues with import